### PR TITLE
Add rudimentary import report.

### PR DESCRIPTION
**Pull request by johndagostino**

### CODE REVIEW

File: lib/dataly/importer.rb [hunk 1]
```diff
@@ -26,25 +26,30 @@ def initialize(filename, options = {})
...
       @default_mapper = options.fetch(:default_mapper, Mapper)
       @default_creator = options.fetch(:default_creator, Dataly::Creator)
+      @import_reporter = Dataly::Reporting::ImportReporter.new(filename)
     end
 
     def process
       csv.each do |row|
         data = mapper.process(row)
         process_create(data)
+        @import_reporter.row_added()
```
Comment (nsinenko) [inline lib/dataly/importer.rb]: don't need `()`

File: lib/dataly/reporting/import_report.rb [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+    def initialize(filename, total_rows, errors)
+      @filename = filename
+      @total_rows = total_rows
+      @errors = errors
+    end
+
+
+    def to_s
+      report = <<"END_OF_REPORT"
+Total rows imported : #{@total_rows}
```
Comment (nsinenko) [inline lib/dataly/reporting/import_report.rb]: You have `attr_reader`, don't need to user instance variables.

File: lib/dataly/reporting/import_report.rb [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+------------------------------------
+
+Errors:
+#{@errors.each { |error| error.message }}
```
Comment (nsinenko) [inline lib/dataly/reporting/import_report.rb]: add line break after each error message?

File: lib/dataly/reporting/import_reporter.rb [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+  def report
+    Dataly::Reporting::ImportReport.new(filename, total_rows, errors)
+  end
+
+  def report_error(error, data = nil)
+    @errors << Dataly::Reporting::RowError.new(error, data)
+  end
+
+  def row_added
+    @total_rows = @total_rows + 1
```
Comment (nsinenko) [inline lib/dataly/reporting/import_reporter.rb]:
``` ruby
attr_accessor :total_rows

def add_row
  total_rows += 1
end
```

File: lib/dataly/reporting/import_reporter.rb [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+    @total_rows = 0
+    @filename = filename
+  end
+
+  def report
+    Dataly::Reporting::ImportReport.new(filename, total_rows, errors)
+  end
+
+  def report_error(error, data = nil)
+    @errors << Dataly::Reporting::RowError.new(error, data)
```
Comment (nsinenko) [inline lib/dataly/reporting/import_reporter.rb]: why define `attr_reader :errors, :total_rows, :filename` and then reference instance variables?

File: lib/dataly/reporting/import_report.rb [hunk 6]
```diff
@@ -0,0 +1,25 @@
+module Dataly::Reporting
+  class ImportReport
+    attr_reader :filename, :errors, :total_rows
+
+    def initialize(filename, total_rows, errors)
+      @filename = filename
+      @total_rows = total_rows
+      @errors = errors
+    end
+
```
Comment (nsinenko) [inline lib/dataly/reporting/import_report.rb]: extra line break

File: lib/dataly/importer.rb [hunk 7]
```diff
@@ -26,25 +26,30 @@ def initialize(filename, options = {})
       @errors = options.fetch(:errors, :log)
       @default_mapper = options.fetch(:default_mapper, Mapper)
```
Comment (nsinenko) [inline lib/dataly/importer.rb]: Can you use named params instead of `options.fetch`?

File: lib/dataly/reporting/import_report.rb [hunk 8]
```diff
@@ -0,0 +1,25 @@
+module Dataly::Reporting
```
Comment (nsinenko) [inline lib/dataly/reporting/import_report.rb]: Why does it need it's own class for a 5-line method? All it does is formatting the message.

File: lib/dataly/importer.rb [hunk 9]
```diff
@@ -26,25 +26,30 @@ def initialize(filename, options = {})
...
         creator.create(data)
       rescue Exception => e
-        create_failed!(e)
+        create_failed!(e, data)
       end
     end
 
-    def create_failed!(e)
+    def create_failed!(e, data)
+      logger.error(data)
```
Comment (nsinenko) [inline lib/dataly/importer.rb]: Add a timestamp to error messages.

File: lib/dataly/reporting/import_report.rb [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+
+    def initialize(filename, total_rows, errors)
+      @filename = filename
+      @total_rows = total_rows
+      @errors = errors
+    end
+
+
+    def to_s
+      report = <<"END_OF_REPORT"
```
Comment (nsinenko) [inline lib/dataly/reporting/import_report.rb]: Is this actually "END OF REPORT" or start? Why use underscores? It's not a variable.

File: lib/dataly/reporting/import_report.rb [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+Total rows imported : #{@total_rows}
+------------------------------------
+
+Errors:
+#{@errors.each { |error| error.message }}
+
+END_OF_REPORT
+      report
```
Comment (nsinenko) [inline lib/dataly/reporting/import_report.rb]: don't need to return

File: lib/dataly/reporting/import_report.rb (hunk 10 above)
Comment (nsinenko) [inline lib/dataly/reporting/import_report.rb]: don't need a variable here

File: lib/dataly/reporting/import_report.rb [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+    end
+
```
Comment (nsinenko) [inline lib/dataly/reporting/import_report.rb]: extra line break

File: lib/dataly/reporting/import_report.rb (hunk 11 above)
Comment (chaitanyakuber) [inline lib/dataly/reporting/import_report.rb]: we should redo this at some stage

File: lib/dataly/reporting/row_error.rb [hunk 13]
```diff
@@ -0,0 +1,9 @@
+class Dataly::Reporting::RowError
```
Comment (nsinenko) [inline lib/dataly/reporting/row_error.rb]: Don't need a new class for this one.

File: lib/dataly/reporting/import_report.rb (hunk 2 above)
Comment (chaitanyakuber) [inline lib/dataly/reporting/import_report.rb]: good point, missed that.

**Comment by chaitanyakuber** [comment]
all minor points, given looming deadline on monday merge away if u need to :+1:

File: lib/dataly/reporting/import_report.rb (hunk 11 above)
Comment (nsinenko) [inline lib/dataly/reporting/import_report.rb]: @chaitanyakuber I just think we should have some degree of QA for open source gems, potential job candidates do look though our public repos.

Comment (chaitanyakuber) [inline lib/dataly/reporting/import_report.rb]: @nsinenko no arguments about quality bar, but the current deadline is important. i am sure the author(s) will come back and clean this up.

**Pull request merged**